### DISCUSSION TITLE
Allow summon typeclass instance of opaque type defined in companion object while deriving.

### DISCUSSION THREAD

**Discussion by jilen** [scala/scala3#17640] [Feature Requests]
```scala
import io.circe._
case class Foo(id: Foo.Id) derives Codec.AsObject // Cannot find encoder/decoder of Id
object Foo {
  opaque type Id = Long
}
```
The above very common opaque  type use case won't compile.
Seems compiler don't know `Id` is `Long` while deriving. 

Note, if opaque type defined in top level, it compiles just fine.
```scala
import io.circe._
opaque type FooId = Long
case class Foo(id: FooId) derives Codec.AsObject // Compiles ok

```

**Comment by ivan-klass** [comment]
I think the one I created might be related: https://github.com/lampepfl/dotty-feature-requests/issues/344

I personally think that type classes for opaque types shouldn't automatically pass-through to underlying type, as it's breaking the concept of a "different type".
Consider the following:
```scala
opaque type PositiveInt = Int
def positive(i: Int): Option[PositiveInt] = Option.when(i > 0)(i)
def toInt(p: PositiveInt): Int = p
```

**Comment by jilen** [comment]
Not pretty sure, but I thought opaque type is basically used as `newtype` instead of `refined` type.

**Comment by ivan-klass** [comment]
> Not pretty sure, but I thought opaque type is basically used as `newtype` instead of `refined` type.

However, `Refined` library is implemented exactly as [opaque types for scala3](https://github.com/fthomas/refined/blob/dca5ece2f0befcdff12d5bdc5c2717bba4e8a6b8/modules/core/shared/src/main/scala-3.0%2B/eu/timepit/refined/api/Refined.scala#L3). And many typeclasses (e.g. `io.circe.Decoder`, `sttp.tapir.Shema`, `com.monovore.decline.Argument` has extra modules to derive their typeclasses properly for `Refined` types, respecting their "business-logic".

In general I'm trying to say that by-passing typeclasses to underlying type shouldn't be automatic as it will make them unsafe - the intention of opaque type as I see it - is to **encapsulate** the underlying type at zero runtime cost. If no encapsulation is cared, then why not using a simple type alias (without opaque keyword)?

But as for companion-object case, I agree that by-passing opaque typeclass for derivation looks eligible, because the opaque type defined right in the companion object which (I believe) is the scope of derivation...
So lampepfl/dotty/issues/17262 is the root cause and the thing to fix
